src/Telomare/Resolver.hs

```diff
@@ -61,6 +59,170 @@ i2c x = TLam (Closed "f") (TLam (Open "x") (inner x))
+  PatternPair x y   -> PairUP (pattern2UPT x) (pattern2UPT y)
+  PatternInt i      -> IntUP i
+  PatternString str -> StringUP str
+  PatternVar str    -> IntUP 0
+  PatternIgnore     -> IntUP 0
+    -- Note that "__ignore" is a special variable name and not accessible to users because
+    -- parsing of VarUPs doesn't allow variable names to start with `_`
+
+mkCaseAlternative :: UnprocessedParsedTerm -- ^ UPT to be cased
+                  -> UnprocessedParsedTerm -- ^ Case result to be made lambda and applied
+                  -> Pattern -- ^ Pattern
+                  -> UnprocessedParsedTerm -- ^ case result as a lambda applied to the appropirate part of the UPT to be cased
+mkCaseAlternative casedUPT caseResult p = appVars2ResultLambdaAlts patternVarsOnUPT . makeLambdas caseResult . keys $ patternVarsOnUPT where
+  patternVarsOnUPT :: Map String UnprocessedParsedTerm
+  patternVarsOnUPT = ($ casedUPT) <$> findPatternVars p
+  appVars2ResultLambdaAlts :: Map String UnprocessedParsedTerm
+                           -> UnprocessedParsedTerm -- ^ case result as lambda
+                           -> UnprocessedParsedTerm
+  appVars2ResultLambdaAlts m = \case
+    lam@(LamUP varName upt) ->
+      case Map.lookup varName m of
+        Nothing -> lam
+        Just x -> AppUP (LamUP varName (appVars2ResultLambdaAlts (Map.delete varName m) upt)) x
+    x -> x
+  makeLambdas :: UnprocessedParsedTerm
+              -> [String]
+              -> UnprocessedParsedTerm
+  makeLambdas upt = \case
+    []     -> upt
+    (x:xs) -> LamUP x $ makeLambdas upt xs
+
+case2annidatedIfs :: UnprocessedParsedTerm -- ^ Term to be pattern matched
+                  -> [Pattern] -- ^ All patterns in a case expression
+                  -> [UnprocessedParsedTerm] -- ^ Int leaves as ListUPs on UPT
+                  -> [UnprocessedParsedTerm] -- ^ Int leaves as ListUPs on pattern
+                  -> [UnprocessedParsedTerm] -- ^ String leaves as ListUPs on UPT
+                  -> [UnprocessedParsedTerm] -- ^ String leaves as ListUPs on pattern
+                  -> [UnprocessedParsedTerm] -- ^ Case's alternatives
+                  -> UnprocessedParsedTerm
+case2annidatedIfs _ [] [] [] [] [] [] =
+  ITEUP (IntUP 1)
+        (AppUP (VarUP "abort") $ StringUP "Non-exhaustive patterns in case")
+        (IntUP 0)
+case2annidatedIfs x (aPattern:as) (dirs2IntsOnUPT:bs) (dirs2IntsOnPattern:cs) (dirs2StringOnUPT:ds) (dirs2StringOnPattern:es) (resultAlternative:fs) =
+  let patternVarsOnUPT :: Map String UnprocessedParsedTerm
+      patternVarsOnUPT = ($ x) <$> findPatternVars aPattern
+  in ITEUP (AppUP (AppUP (AppUP (VarUP "foldl")
+                                 (VarUP "and"))
+                         (IntUP 1))
+                  (ListUP [ AppUP (AppUP (VarUP "listEqual") dirs2IntsOnUPT) dirs2IntsOnPattern
+                          , AppUP (AppUP (VarUP "listEqual") dirs2StringOnUPT) dirs2StringOnPattern
+                          , pairStructureCheck aPattern x
```

Comment:
This seems like it will generate some overhead for performing tests for int/string matches where there aren't ints/strings. This wouldn't be a big deal, if Telomare wasn't so slow at the moment :( (I'm making progress on that, I swear!)

Reply:
Implemented a version with less overhead. let me know what you think